### ISSUE TITLE
Use semantic release in metac

### ISSUE THREAD

**Issue by grzesuav** [AmitKumarDas/metac#64]
To make automatic release notes and make easier to deploy new version, `semantic-release` could be used. https://github.com/semantic-release/semantic-release is a tool which :
* create release notes from commit messages
* creates releases on gihub  https://github.com/AmitKumarDas/metac/releases page
* version the release automatically

It would need :
* make commit messages follow some conventions (which can be problematic it they weren't any )

**Comment by grzesuav** [comment]
@AmitKumarDas  I can try to add support (I have done it in my gitlab project ) probably shouldn't be hard for github.

**Comment by AmitKumarDas** [comment]
Sounds great @grzesuav

**Comment by grzesuav** [comment]
briefly looking on commit messages: 

```
8742540 [example] update steps to automate uninstall openebs (#62)
a238880 [example] update uninstall openebs example to latest metac (#61)
4ea7d56 [gctl] add UpdateDuringPendingDelete option (#60)
9af1eb8 [example] fix example on uninstall openebs (#59)
00b8375 [example] update rbac for uninstall-openebs example (#58)
a74a6fe [fix] name of finalizer assigned to watch resource (#57)
df70117 [example] update uninstall-openebs to latest metac (#56)
ee6f735 [config] fix loading of metac config file(s) (#55)
e6c273c [example] add updated metac to clean uninstall openebs (#54)
fd7e447 [example] add example to clean uninstall openebs (#53)
c624ab1 [readme] update readme with latest info about metac (#52)
3a0c517 [select] add advanced select rules to match specific resource(s) (#51)
59073f6 [config] add wait for all metac configs to load (#50)
ea19971 [example] improved README for set-status-on-cr example (#49)
cb54722 [example] update set-status-on-cr example with README.md & test.sh (#48)
ef8f6d5 [example] update set-status-on-cr to latest metac commit (#47)
0065a1c [fix] running metac from config (#46)
89bf818 [example] update set-status-on-cr to latest metac commit (#45)
d6ba051 [fix] config load by adding path along with file name (#44)
5c9a91b [example] update set-status-on-cr to latest metac commit (#43)
1c3c1a4 [local] fix config to avoid directories (#42)
eac4e63 [example] changes to set status on cr to point to latest metac (#41)
5df534c [local] add logs to local config based metac run (#40)
768f916 [example] set status on custom resource via inline hook (#39)
7c5d651 [start] move start code from main.go to start package (#38)
e24638c add generated client code to let metac be used as a lib (#37)
b625ffc [hook] add support to invoke inline functions as hooks (#36)
fb35f12 separate selector logic for watch & attachments of gctl (#35)
31c15db add skipReconcile to GenericController hook response (#34)
873ada4 [local] add option to run metac from local config (#33)
806e3a2 Adding a notice to the README about finalizers (#31)
d910577 adding a helmchart to deploy metac (#30)
6383e2c Make client-go throttling configurable. (#29)
41435e8 (tag: v0.1.0) [usecase, gctl]: showcase install/un-install of CRD (#28)

```
so the closes match is https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-ember#ember-convention . 

However I often find used Angular commit conventions (https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#type) (at least up to types) which look likes :
```
chore(github): update Issue template with LTS info
```

I would propose the similar without the scope (int the `(` `)` brackets) :

```
chore: #555 - Update Issue template with LTS info
```

but it should be discussed and enforced in github checking for agreed commit messages format (like via https://github.com/probot/semantic-pull-requests or by some travis job check)

Commits up to now can be handled differently (like do not rewrite history, just map them somehow to major/minor/path/chore) and enforce format till this moment.

@AmitKumarDas any opinions on that ?

**Comment by AmitKumarDas** [comment]
`chore: #555 - Update Issue template with LTS info`

I guess above looks fine. I hope #555 is issue number. Issue may be optional perhaps.

**Comment by AmitKumarDas** [comment]
Can I set the old commits with labels like major, minor, chore etc. Will that do without getting too much into git stuff.

**Comment by AmitKumarDas** [comment]
@kmova can you please provide your thoughts as well.

**Comment by grzesuav** [comment]
> Can I set the old commits with GitHub labels like major, minor, chore etc. Will that do without getting too much into git stuff.

No, I am afraid not. Semantic release relies only on git commit messages. Other  option is to `tag` new release manually and enforce standards and introduce semantic release since this moment.

**Comment by grzesuav** [comment]
Moreover (https://github.com/semantic-release/semantic-release/blob/master/docs/support/FAQ.md#can-i-set-the-initial-release-version-of-my-package-to-001) they recommend to start from `1.0.0` so maybe it would be possible to manually tag a `1.0.0` with release notes etc and since then rely on commit messages ?


By the way @AmitKumarDas I have some questions :
1. Which CI is used for doing build ? I do not mean test but actually to provide binary packages/docker  images
2. What are build artifacts ? Go binary/docker image ?

**Comment by kmova** [comment]
@grzesuav - thanks for bringing up these questions. 

I agree we should go ahead and make a 1.0.0 release and start enforcing the commit message format. 

- Regarding the commit message format, can it be:
  ```
  type(module): commit description specifying what is done #issue-number
  ```
  At times, a PR can be raised without an issue, so the #issue number can be optional. 

- For CI, we could use travis, but open to ideas. 

Also, is there a way to enforce commit message format in github?

**Comment by grzesuav** [comment]
@kmova - yes, for format message it is actually an angular message format - https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines . I would say that (scope) should be also optional - actually we only need `type` to make semantic-release work

for enforcing commit messages for example https://probot.github.io/apps/commitlint/


so, travis is also building go binaries and making and push image to docker registry ?

**Comment by AmitKumarDas** [comment]
@grzesuav Triggers have been set at [quay](https://quay.io/repository/amitkumardas/metac) to build docker images of metac whenever any PR is pushed against [metac repo](https://github.com/AmitKumarDas/metac/). These images are hosted at quay _(similar to dockerhub)_.

One needs to use `docker pull quay.io/amitkumardas/metac` or `quay.io/amitkumardas/metac` as the image name. One can also use various tags present [here](https://quay.io/repository/amitkumardas/metac?tab=tags)

**Comment by grzesuav** [comment]
@AmitKumarDas ok, so on `tag` there will be `tagged` docker release made there ? Likes `quay.io/amitkumardas/metac:1.0.0` ?

**Comment by AmitKumarDas** [comment]
Yes. @grzesuav 
It's the same case when we cut a GitHub release on the repo. This release version is automatically reflected as a tag in the docker image.

**Comment by grzesuav** [comment]
hi @AmitKumarDas , maybe we can push this further ? We need to :
* tag release 1.0.0 manually
* add commit verificator (to check commit prefixes)

and then I can add semantic release (and automate relasign basing on it)

**Comment by AmitKumarDas** [comment]
Hi @grzesuav 
Do we need to first tag a 1.0.0 & then add the commit verification?

Do you have in mind any commit verificator that can be added to this repo right away? I can add the same or even accept any PR?  

I shall do the manual release of 1.0.0 shortly.
cc @luisdavim do you have any reservations on tagging.

**Comment by grzesuav** [comment]
Probably can add commit verification earlier, could try

**Comment by luisdavim** [comment]
We should also start publishing the helm chart to https://hub.helm.sh/ using the same version as the docker tag.

**Comment by grzesuav** [comment]
Initial config https://github.com/AmitKumarDas/metac/pull/104

**Comment by grzesuav** [comment]
@luisdavim yes, version should be sourced from git-tag. Release in semantic-release trigger new tag, which pushes docker image and helm chart release. Not sure @AmitKumarDas you are also publishing go package somewhere ?

**Comment by grzesuav** [comment]
after merging (or before) one of the maintainers should install https://github.com/apps/semantic-pull-requests - @AmitKumarDas can you do it ? 

Configuration (types - like feat/fix/chore/etc) can be adjusted later (I will try to add semantic-release and config in one go)

**Comment by grzesuav** [comment]
For semantic release itself, it can be done via :
* github action (like https://github.com/marketplace/actions/action-for-semantic-release)
* travis ci

what is your preference ?

**Comment by AmitKumarDas** [comment]
> For semantic release itself, it can be done via :
> github action (like https://github.com/marketplace/actions/action-for-semantic-release)
> travis ci
> what is your preference ?

Let us try with github actions.
I have added NPM_TOKEN & GITHUB_TOKEN as personal access tokens. However, I am not sure of the level of access that needs to be granted.

**Comment by AmitKumarDas** [comment]
> after merging (or before) one of the maintainers should install https://github.com/apps/semantic-pull-> requests - @AmitKumarDas can you do it ?

Done with merge as well as installing above app!

**Comment by grzesuav** [comment]
@AmitKumarDas https://github.com/AmitKumarDas/metac/pull/105

![image](https://user-images.githubusercontent.com/9568421/76450284-cc953f00-63cd-11ea-8f26-fd42a017ae25.png)


so seems working

**Comment by grzesuav** [comment]
I have played a bit, amending commit and editing PR title to semantic form makes it unblock the PR

**Comment by grzesuav** [comment]
https://github.com/AmitKumarDas/metac/pull/109

**Comment by grzesuav** [comment]
hi, example config there (and result of running on my fork). some remarks :
1. every push to master will trigger a release 
2. currently tests are run on travis and independently github action triggers a release. Therefore master could fail and release could be trigered. Mitigation needed.
3. release creates tag. Need to add triggering a docker build on tag. @AmitKumarDas how it is done currently ? I would imagine another github action can be added on tag to trigger releasing a docker image with correct version
4. build artifacts - are they any ? Should we publish something when triggering a release, a binary or sth ? [I would imagine it would be useful]
5. having changelog plugin would allow to have `CHANGELOG.MD` file (name could be customized) maintanable


any other suggestions ?

**Comment by grzesuav** [comment]
around 2. - possible solutions :
* wrap running travis ci into workflow and trigger release on master only when travis ci passed
* run tests as github action

**Comment by grzesuav** [comment]
around 3. 
* I see advantages of keeping whole CI in one place , looking at https://quay.io/repository/amitkumardas/metac?tab=tags it creates a lot of versions

**Comment by AmitKumarDas** [comment]
@grzesuav Any CI or github actions that makes the entire workflow smooth will be good.
I would need help from your side to setup these.

Travis CI is used to check if integration test, unit test & docker image build is actually passing.
Quay is setup from my own repository/login after integrating it with metac repo. Quay triggers build for every PR & merge to metac.
Release is a manual tagging process that asks for a release version & upon setting triggers a build which in turn triggers Travis & Quay.

**Comment by grzesuav** [comment]
sure, I can do it. But wanted to confirm flow first :
1. Run test via github actions on every push/PR- to check if it is possible
2. When test pass on master - trigger a semantic-release, which creates also a tag
3. When on tag - publish artifacts (if possible) and trigger docker push

@AmitKumarDas are you ok with that ?

**Comment by grzesuav** [comment]
I have found https://github.com/marketplace/actions/build-tag-publish-docker for building containers in github action

**Comment by grzesuav** [comment]
@AmitKumarDas I have integrated testing and releasing on single workflow :
example : https://github.com/grzesuav/metac/actions/runs/58455222

tests and integrations-test are run in parallel, when both finish successfully, then (if on master branch) semantic-release will trigger a release

**Comment by grzesuav** [comment]
I believe now we can switch from quay-github integration and move build & push docker also to gitlab action.



> @grzesuav Any CI or github actions that makes the entire workflow smooth will be good.
> I would need help from your side to setup these.
> 
> Travis CI is used to check if integration test, unit test & docker image build is actually passing.
> Quay is setup from my own repository/login after integrating it with metac repo. Quay triggers build for every PR & merge to metac.
> Release is a manual tagging process that asks for a release version & upon setting triggers a build which in turn triggers Travis & Quay.

So this will need to change as :
* currently sem-rel is set to trigger release on every merge to master (if there will be `fix`/`feat` commit)
* docker images should be pushed only for releases

**Comment by grzesuav** [comment]
There is no manual action type on github, closest I could find would be repository dispatch event if I read this correctly : https://help.github.com/en/actions/reference/events-that-trigger-workflows

**Comment by AmitKumarDas** [comment]
> sure, I can do it. But wanted to confirm flow first :
> 
> 1. Run test via github actions on every push/PR- to check if it is possible
> 2. When test pass on master - trigger a semantic-release, which creates also a tag
> 3. When on tag - publish artifacts (if possible) and trigger docker push
> 
> @AmitKumarDas are you ok with that ?

Yes

**Comment by grzesuav** [comment]
Found official https://github.com/marketplace/actions/build-and-push-docker-images#Example-usage docker plugin.

@AmitKumarDas could you encrypt in gitlab `DOCKER_USERNAME` and `DOCKER_PASSWORD` variables ? 

Also, what is registry url for quay.io ?

**Comment by grzesuav** [comment]
Added changes to https://github.com/AmitKumarDas/metac/pull/109, however not sure how to test docker image push

**Comment by AmitKumarDas** [comment]
@grzesuav  https://quay.io/repository/amitkumardas/metac

Should we push to some docker registry?
We can create a new registry?

**Comment by grzesuav** [comment]
> @grzesuav https://quay.io/repository/amitkumardas/metac
> 
> Should we push to some docker registry?
Docker images must be in docker registry. I believe quay.io is docker registry

> We can create a new registry?

It would be possible, although usually convention is to have repository/project so I think it could stay like it is. 

As for me can try to use docker registry in form I configure it, and we will see how it works. I ca adjust it later with `fix: ` commit which would just trigger minor release.

What do you think ?

**Comment by AmitKumarDas** [comment]
sounds great @grzesuav 
I am merging the PR.

**Comment by AmitKumarDas** [comment]
@grzesuav i am not sure of the credentials that need to be set against following ENVs
```
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
```
I will look around to find the clues

**Comment by AmitKumarDas** [comment]
@grzesuav release got created.
More additions are needed to following file:
https://github.com/AmitKumarDas/metac/blob/master/.github/semantic.yml

For example more types needs to be added:
https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#type

**Comment by grzesuav** [comment]
@AmitKumarDas fix for docker release https://github.com/AmitKumarDas/metac/pull/116

seems separate workflow is needed to be triggered on tags

**Comment by grzesuav** [comment]
> @grzesuav i am not sure of the credentials that need to be set against following ENVs
> 
> ```
>           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
>           NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
> ```
> 
> I will look around to find the clues

GITHUB token is needed for semantic release to communicate with github API and :
* create a tag
* create a release
* make comments on PR/issues

NPM is not needed, it is for publishing for nmp repository, I will remove it  in separate PR

**Comment by grzesuav** [comment]
> @grzesuav a new metac release got created.
> More additions are needed to following file:
> https://github.com/AmitKumarDas/metac/blob/master/.github/semantic.yml
> 
> For example more types needs to be added:
> https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#type

Yes, will look on that. Need to have same config for PR watcher and semantic release itself.

**Comment by luisdavim** [comment]
For tests I've started using these recently: 
- https://github.com/bats-core/bats-core
- https://github.com/vincent-zurczak/detik
- https://github.com/jasonkarns/bats-mock

**Comment by AmitKumarDas** [comment]
> For tests I've started using these recently:
> 
> https://github.com/bats-core/bats-core
> https://github.com/vincent-zurczak/detik
> https://github.com/jasonkarns/bats-mock

Hi @luisdavim above looks great. If we can integrate one of above with metac/examples & then run that on some K8s env (KIND or minikube) inside Travis or github actions then we achieve an automated e2e. Was that the intent of above post?

Meanwhile, I am working on a project around declarative e2e testing using metac. I shall talk about it once I am done.

**Comment by luisdavim** [comment]
I've been using `k3d`. What I have at the moment is very specific to the setup we use at my workplace, I'll see if I get some free time over the weekend and create something specific to `metac`.

**Comment by grzesuav** [comment]
hmm noting happened. I was testing it by making and pushing my tags, probably when tag is created on gitlab it happens without pushing ? Wil try to cover it today with different workflow enabler.

@luisdavim does bats is for tasting bash scripts itself ?

**Comment by grzesuav** [comment]
@AmitKumarDas finally found it ! https://github.community/t5/GitHub-Actions/Github-actions-workflow-not-triggering-with-tag-push/td-p/39685 it seems that Workflow is not triggered by events created by default `GITHUB_TOKEN`. As a workaround - could you set secret named `PAT` on the repository (personal access token). It requires only repo permissions :
![image](https://user-images.githubusercontent.com/9568421/77804323-899ac300-707f-11ea-919f-ca2e9d35b5c4.png)
I have made https://github.com/AmitKumarDas/metac/pull/117 to address this

**Comment by AmitKumarDas** [comment]
@grzesuav I did that & merged the PR later. However, git push seems to fail due to insufficient permissions. I am still trying to find the exact permissions in addition to the ones mentioned above.

**Comment by AmitKumarDas** [comment]
Latest pr passed fine.. not sure if things are fine now.

**Comment by AmitKumarDas** [comment]
@grzesuav Further updates. Submitting a PR passes with whatever credentials we have. However, post merging a PR, the release process fails.

**Comment by grzesuav** [comment]
> @grzesuav Further updates. Submitting a PR passes with whatever credentials we have. However, post merging a PR, the release process fails.

Don't look at PR status of Release job - it will skip releasing as not on master branch : 

`[1:42:33 PM] [semantic-release] › ℹ  This test run was triggered on the branch code-2, while semantic-release is configured to only publish from master, therefore a new version won’t be published.`


At master there is :
'**semantic-release** cannot push the version tag to the branch `master` on the remote Git repository with URL `***github.com/AmitKumarDas/metac`.\n' +
    '\n' +
    'This can be caused by:\n' +
    ' - a misconfiguration of the [repositoryUrl](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#repositoryurl) option\n' +
    ' - the repository being unavailable\n' +
    ' - or missing push permission for the user configured via the [Git credentials on your CI environment](https://github.com/semantic-release/semantic-release/blob/master/docs/usage/ci-configuration.md#authentication)'

**Comment by grzesuav** [comment]
> @grzesuav a new metac release got created.
> More additions are needed to following file:
> https://github.com/AmitKumarDas/metac/blob/master/.github/semantic.yml
> 
> For example more types needs to be added:
> https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#type

https://github.com/zeke/semantic-pull-requests#configuration by default all of those are supported and accepted. Or you have some example it didn't worked ?

**Comment by AmitKumarDas** [comment]
If default works then great. I thought we have to set it.

**Comment by grzesuav** [comment]
> If default works then great. I thought we have to set it.

Default actually **does not work**, unfortunately as expected. I added few additions in https://github.com/AmitKumarDas/metac/pull/120 . 

How it should work :
* after merge to master, release 0.2.2 should be created
* `Docker release` workflow should be triggered

using this PR, on my repository :
![image](https://user-images.githubusercontent.com/9568421/77950459-d0c7c480-72c8-11ea-80c4-50f0552180e3.png)
* Test and release is triggered on branch `try_gh_token`
* After merging to master, it is triggered again, on `master` branch and this time it triggers a release, tag is created
* On tag creation, workflow `Release docker  image` is triggered (it fails as I do not have account on quay to push)

**Comment by AmitKumarDas** [comment]
@grzesuav Can you try any of the following secrets as reqd:
 DOCKER_PASSWORD	
 DOCKER_USERNAME	
 PAT

**Comment by grzesuav** [comment]
@AmitKumarDas do not now why but it jus started working :
![image](https://user-images.githubusercontent.com/9568421/78047416-987fbf00-7378-11ea-93ff-1330839a4e67.png)
and we have now release 0.2.2 :
![image](https://user-images.githubusercontent.com/9568421/78047476-adf4e900-7378-11ea-9fa1-80f6c147667b.png)
it happend 12 hours ago (more or less when you were writing previous comment)

**Comment by grzesuav** [comment]
My suggestions :
* we  can disable travis 
* quay.io image build - not it is triggered by ever push to repo. Also the image is pushed from tag. It 
 overlaps as can be seen here - https://quay.io/repository/amitkumardas/metac?tab=history
 
 ![image](https://user-images.githubusercontent.com/9568421/78048406-f7920380-7379-11ea-864b-3dbd1ab24d8b.png)

 so first image was build and push by github, and then quay.io triggered second build on the same 
 tag and override tag. 

I would suggesting  disabling quay.io builds, as it is covered by github. 

* it was suggested to make helm chart and integrate it, maybe a separate issue for that ? 

@AmitKumarDas any opinions on above ?

**Comment by AmitKumarDas** [comment]
We will disable Travis once we are happy with GitHub actions. 
IMO let quay be a duplicate registry unless it creates confusion for end users.
I am not sure if I understand integration with helm. @luisdavim can you throw some light on helm integration? What exactly should we do?
